thewalrus/fock_gradients.py

```diff
@@ -237,111 +307,74 @@ def grad_BSgate(T, gradTtheta, gradTphi, phi):  # pragma: no cover
+
+    # rank 3
+    for m in range(0, cutoff):
+        for n in range(0, cutoff - m):
+            p = m + n
+            if 0 < p < cutoff:
+                Z[m, n, p, 0] = (
+                    R[0, 2] * sqrt[m] / sqrt[p] * Z[m - 1, n, p - 1, 0]
+                    + R[1, 2] * sqrt[n] / sqrt[p] * Z[m, n - 1, p - 1, 0]
+                )
+
+    # rank 4
+    for m in range(0, cutoff):
+        for n in range(0, cutoff):
+            for p in range(0, cutoff):
+                q = m + n - p
+                if 0 < q < cutoff:
+                    Z[m, n, p, q] = (
+                        R[0, 3] * sqrt[m] / sqrt[q] * Z[m - 1, n, p, q - 1]
+                        + R[1, 3] * sqrt[n] / sqrt[q] * Z[m, n - 1, p, q - 1]
+                    )
 
+    return Z[:cutoff, :cutoff, :cutoff, :cutoff]
 
-@jit("void(double[:,:], double[:,:])")
-def grad_Sgate_real(T, gradT):  # pragma: no cover
-    """Calculates the gradient of the Sgate.
-
-    Args:
-        T (array[float]): array representing the gate
-        gradT (array[float]): array of zeros that will contain the value of the gradient
-    """
-    cutoff = gradT.shape[0]
-    for n in range(cutoff):
-        offset = n % 2
-        for m in range(offset, cutoff, 2):
-            gradT[n, m] = -0.5 * np.sqrt((m + 1) * (m + 2)) * T[n, m + 2]
-            if m > 1:
-                gradT[n, m] += 0.5 * np.sqrt(m * (m - 1)) * T[n, m - 2]
 
-
-def Sgate_real(s, cutoff, grad=False):
-    """Calculates the Fock representation of the Sgate and its gradient.
+def BSgate(theta, phi, cutoff, grad=False):
+    r"""Calculates the Fock representation of the S2gate and its gradient.
 
     Args:
-        s (float): parameter of the gate
+        theta (float): transmissivity angle of the beamsplitter. The transmissivity is :math:`t=\cos(\theta)`
+        phi (float): reflection phase of the beamsplitter
         cutoff (int): Fock ladder cutoff
         grad (boolean): whether to calculate the gradient or not
 
     Returns:
-        tuple[array[float], array[float] or None]: The Fock representations of the gate and its gradient with size ``[cutoff]*2``
+        tuple[array[float], array[float] or None]: The Fock representations of the gate and its gradient with size ``[cutoff]*4``
     """
-    mat = np.array([[np.tanh(s), -1.0 / np.cosh(s)], [-1.0 / np.cosh(s), -np.tanh(s)]])
     if not grad:
-        return squeezing_real(mat, cutoff), None
+        return beamsplitter_rec(theta, phi, cutoff), None, None
 
-    T = squeezing_real(mat, cutoff + 2)
-    gradT = np.zeros([cutoff, cutoff], dtype=float)
-    grad_Sgate_real(T, gradT)
+    T = beamsplitter_rec(theta, phi, cutoff + 1)
+    gradTtheta = np.zeros([cutoff, cutoff, cutoff, cutoff], dtype=complex)
+    gradTphi = np.zeros([cutoff, cutoff, cutoff, cutoff], dtype=complex)
+    grad_BSgate(T, gradTtheta, gradTphi, phi)
 
-    return T[0:cutoff, 0:cutoff], gradT
+    return T[:cutoff, :cutoff, :cutoff, :cutoff], gradTtheta, gradTphi
```

Comment:
Might not be relevant to this PR or not, but wondering if the returned dimensions now match SF?

Reply:
Will also add an sf ordering optional argument.

Reply:
Done!